mgmt: mcumgr: Add ability for MCUmgr to setup at boot-up

nordicjm opened the pull request:
Adds an optional Kconfig that allows MCUmgr to set itself up at
    boot-up, register command handlers and initiate transports,
    instead of needing user-application interaction to perform this.

Fixes #52154

**Note** First commit is from https://github.com/zephyrproject-rtos/zephyr/pull/52356 and can be ignored

aunsbjerg commented on subsys/mgmt/mcumgr/transport/smp_udp.c:
```diff
@@ -279,3 +290,31 @@ int smp_udp_close(void)
+			LOG_ERR("Could not open SMP UDP");
+		}
+
+		return;
```
Both the explicit return statements here are redundant and can be removed.

nordicjm commented on subsys/mgmt/mcumgr/transport/smp_udp.c:
```diff
@@ -279,3 +290,31 @@ int smp_udp_close(void)
+	}
+}
+
+void smp_udp_boot_init(void)
```
Fixed

nordicjm commented on include/zephyr/mgmt/mcumgr/smp_udp.h:
```diff
@@ -17,19 +17,24 @@ extern "C" {
+/**
+ * @brief       Starts UDP transport at start-up if ``CONFIG_MCUMGR_SMP_BOOT_SETUP`` is enabled.
+ */
+void start_smp_bt(void);
```
Fixed

de-nordic commented:
The Zephyr group already does automatic registration, but uses different approach:
https://github.com/zephyrproject-rtos/zephyr/blob/b49b11555c9ae33a22e3cc0586a0ae31842507f2/subsys/mgmt/mcumgr/lib/cmd/zephyr_basic/src/basic_mgmt.c#L61-L70

nordicjm commented:
> The Zephyr group already does automatic registration, but uses different approach:

Question is, should this be centralised so that all commands are registered at boot (and have a Kconfig for the transports to set them up automatically) using a single SYS_INIT() rather than multiple, or go with multiple?

nordicjm commented:
Updated to move the zephyr basic management setup to the new system

nordicjm committed:
mgmt: mcumgr: Add ability for MCUmgr to setup at boot-up

Adds an optional Kconfig that allows MCUmgr to set itself up at
boot-up, register command handlers and initiate transports,
instead of needing user-application interaction to perform this.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
samples: subsys: mgmt: mcumgr: Update sample to use boot-up Kconfig

This replaces the start up code that registers the command handlers
and transports with the new Kconfig that does this automatically at
boot-up.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
